Develop

digitalsadhu opened the pull request:
Adds a CSS writer to roughly mirror `asset-pipe-js-writer`

Implements simple css bundling using `postcss` to resolve and inline `@import` statements for each css entry point supplied to the writer's constructor.

Stream emits feed items of the form:
```js
{
  id: <hash of name, version and file>,
  name: <module name from closest package.json to file>,
  version: <module version from closest package.json to file>,
  file: <path to file relative to module root (including module name)>,
  content: <bundled css>
}
```

TODO:
- [x] Fix PR comments
- [x] Add snapshots to error message tests

Commit:
Add writer implementation and tests

Commit:
Update to improve writer interface

Add support for a single string file path instead of array
Throw useful error if relative path given to file

SimenB commented on lib/util.js:
```diff
@@ -0,0 +1,38 @@
+};
+
+module.exports.bundleCssModule = function bundleCssModule (filePath) {
+    return readFile(filePath, 'utf8').then(css =>
```
async await here as well :D

trygve-lie commented:
Please add doc to the README from the beginning.

digitalsadhu commented on lib/writer.js:
```diff
@@ -0,0 +1,40 @@
+                this.push(meta);
+            })
+            .catch(err => {
+                process.nextTick(() => this.emit('error', err));
```
Copypasta actually. #willremove

digitalsadhu commented on lib/util.js:
```diff
@@ -0,0 +1,38 @@
+const atImport = require('postcss-import');
+
+module.exports.identifyCssModule = function identifyCssModule (filePath) {
+    return readPkgUp({
```
I found myself trapped between do and do not. Are we supporting latest node only?

SimenB replied:
Well, we are in Podium. Unsure what @trygve-lie wants for the asset-pipe things, but I see no reason to not just support node 8. It will be LTS later this month anyways

digitalsadhu replied:
I’ll take that, run with it and not look back. 
Here comes the async/await train toot toot!

digitalsadhu commented on test/writer.test.js:
```diff
@@ -0,0 +1,189 @@
+});
+
+test('new Writer(filePath)', () => {
+    const filePath = path.join(__dirname, 'test-assets/my-module-1/main.css');
```
Oh shooooot forgot about that. Chur!

digitalsadhu replied:
Oh man, there were actually 3 tests that were not being waited for (with done callbacks or returned promises)

Commit:
Fixes from PR feedback

Commit:
Clean up dependencies

Several deps were unneeded. Removed.

SimenB commented:
You have to add `ecmaVersion: 2017` to `.eslintrc`

Commit:
Fix lint issues

Commit:
Update scripts, remove autolinting before tests

Commit:
Add missing test assets

Commit:
Add assertion expectations, fix tests

Commit:
Pimp out README.md

Commit:
Increase test coverage to 100%, ensure it stays.

Commit:
Add .eslintignore file to prevent linting coverage

Commit:
Fix angry eslint

digitalsadhu commented:
@trygve-lie @SimenB 
I've added README.md content. Fixed PR issues. Increased test coverage to 100% and made it so the tests fail if it drops below 100% coverage.

SimenB commented:
>made it so the tests fail if it drops below 100% coverage.

I ❤️  that

SimenB commented on test/writer.test.js:
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-/* global test, expect, beforeEach */
+/* global test, expect, beforeEach, jest */
```
you can do `env: { jest: true }` in eslintrc

SimenB commented on README.md:
````diff
@@ -1,2 +1,80 @@
+
+```js
+{
+    // Unique id for entry. Created by hashing together name, version and file
````
not the content of the file?

digitalsadhu replied:
hmm, yea I guess I was operating on the assumption that id,name,version,file would be enough since they should be A. unique and B. would change if the package was updated so to include content as well would perhaps handle the case where if somehow the package was updated without updating the version. 
I'll change it to include content as well

SimenB replied:
The use case is e.g. linking some files together locally, I suppose.

The js writer only looks at content. Should those implementations be aligned?

https://github.com/asset-pipe/asset-pipe-js-writer/blob/dac298c19b285a8ae2a64d76a8c894bdc198d773/lib/writer.js#L20

SimenB commented on lib/writer.js:
```diff
@@ -0,0 +1,40 @@
+
+        assert(
+            Array.isArray(files) || typeof files === 'string',
+            `Expected 'files' to be of type 'Array' or a 'string', instead got '${typeof this.files}'`
```
`typeof files`, `this.files` is undefined at this point

digitalsadhu replied:
baam, thats what I get for not asserting on the error message

SimenB replied:
you can to `toThrowErrorMatchingSnapshot()` if you're feeling hipster (it checks the message, not the stack)

digitalsadhu replied:
ah nice, will look into it. Need to increase my Jest knowledge before Thursday so good to check such things out.

SimenB replied:
Feel free to fire questions at me, I'm a collaborator on jest

SimenB commented on package.json:
```diff
@@ -1,52 +1,54 @@
+    },
+    "scripts": {
+        "format":
+            "prettier --write --single-quote --tab-width=4 lib/**/*.js test/**/*.js",
```
`--trailing-comma all` as well

SimenB commented on test/writer.test.js:
```diff
@@ -0,0 +1,248 @@
+'use strict';
+
+/* global test, expect, beforeEach, jest */
```
you can do `env: { jest: true }` in eslintrc (my other comment about the same thing got lost in the eternal void of the internets)

digitalsadhu replied:
Did not know this, cheers. However, the only thing i've found in the past is that sometimes editor plugins don't always respect these sorts of global overrides when not in the same file thats being linted (ie. in package.json or wherever). Other than that i'd prefer not to have to litter the tops of files with these sorts of comments so i'll experiment and see how it goes.

digitalsadhu commented:
Kick ass! Thanks for the review @SimenB Super helpful :)

SimenB commented on .eslintignore:
```diff
@@ -0,0 +1,9 @@
+.idea/**/*
+*.iml
+*.log
+coverage
```
FWIW, only `coverage/` is needed in this file

digitalsadhu replied:
noted. will change

Commit:
Add snapshot testing for error messages

Commit:
Update to include css content in id hash

digitalsadhu committed:
Fixes from PR feedback

SimenB reviewed (approved):
This is an awesome PR. Great job!

SimenB commented on .eslintrc:
```diff
@@ -1,7 +1,11 @@
+    },
+    "extends": ["finn", "finn/node"],
+    "env": {
+        "node": true,
```
`finn/node` sets this. doesn't hurt, of course
